pkg/ottl/parser_collection.go

```diff
@@ -0,0 +1,328 @@
+}
+
+// ottlParserWrapper wraps an ottl.Parser using reflection, so it can invoke exported
+// methods without knowing its generic type (transform context).
+type ottlParserWrapper struct {
+	parser                         reflect.Value
+	prependContextToStatementPaths func(context string, statement string) (string, error)
+}
+
+func newParserWrapper[K any](parser *Parser[K]) *ottlParserWrapper {
+	return &ottlParserWrapper{
+		parser:                         reflect.ValueOf(parser),
+		prependContextToStatementPaths: parser.prependContextToStatementPaths,
+	}
+}
+
+func (g *ottlParserWrapper) parseStatements(statements []string) (reflect.Value, error) {
+	method := g.parser.MethodByName("ParseStatements")
+	parseStatementsRes := method.Call([]reflect.Value{reflect.ValueOf(statements)})
+	err := parseStatementsRes[1]
+	if !err.IsNil() {
+		return reflect.Value{}, err.Interface().(error)
+	}
+	return parseStatementsRes[0], nil
+}
+
+func (g *ottlParserWrapper) prependContextToStatementsPaths(context string, statements []string) ([]string, error) {
+	result := make([]string, 0, len(statements))
+	for _, s := range statements {
+		prependedStatement, err := g.prependContextToStatementPaths(context, s)
+		if err != nil {
+			return nil, err
+		}
+		result = append(result, prependedStatement)
+	}
+	return result, nil
+}
+
+// statementsConverterWrapper is a reflection-based wrapper to the ParsedStatementConverter function,
+// which does not require knowing all generic parameters to be called.
+type statementsConverterWrapper reflect.Value
+
+func newStatementsConverterWrapper[K any, R any](converter ParsedStatementConverter[K, R]) statementsConverterWrapper {
+	return statementsConverterWrapper(reflect.ValueOf(converter))
+}
+
+func (s statementsConverterWrapper) call(
+	parserCollection reflect.Value,
+	ottlParser *ottlParserWrapper,
+	context string,
+	statements StatementsGetter,
+	parsedStatements reflect.Value,
+) (reflect.Value, error) {
+	result := reflect.Value(s).Call([]reflect.Value{
+		parserCollection,
+		ottlParser.parser,
+		reflect.ValueOf(context),
+		reflect.ValueOf(statements),
+		parsedStatements,
+	})
+
+	resultValue := result[0]
+	resultError := result[1]
+	if !resultError.IsNil() {
+		return reflect.Value{}, resultError.Interface().(error)
+	}
+
+	return resultValue, nil
+}
+
+// parserCollectionParser holds an ottlParserWrapper and its respectively
+// statementsConverter function.
+type parserCollectionParser struct {
+	ottlParser          *ottlParserWrapper
+	statementsConverter statementsConverterWrapper
+}
+
+// ParserCollection is a configurable set of ottl.Parser that can handle multiple OTTL contexts
+// parsings, inferring the context, choosing the right parser for the given statements, and
+// transforming the parsed ottl.Statement[K] slice into a common result of type R.
+//
+// Experimental: *NOTE* this API is subject to change or removal in the future.
+type ParserCollection[R any] struct {
```

Comment:
Is there any type of restriction we can put on R? It doesn't feel like a `ottl.Statement[K] slice` could be turned into any other type. In the draft it looks like R is a `consumer.[Traces|Metrics|Logs]`, it would be nice to enforce R's type more if we can.

Reply:
Not sure if we can easily implement it, I can give it a try, but I'm wondering which type restriction would fit here, considering the parser collection is dynamic and can be used to parse any type of context.

The advantage of having it as `any` is flexibility, as consumers could use any custom converter/aggregators on the converter functions, including using the raw `ottl.Statement[K] ` values and `switch r.(type) { case []*Statement[ottllog.TransformContext]: ... }` it later for processing, for example. 

If we don't have any major concern on not restricting it, I'd keep it as `any`, but I'm fine giving it a try if you think otherwise.  Thanks!

Reply:
It would be really great to do something to ensure `return convertedStatements.Interface().(R), nil` doesn't panic.  The `tCtx` (K) being an `any` ends up working out bc the type is only passed around, we never assert anything on it. But with R we attempt to do a cast.  If the `convertedStatements` can't be cast into R we'll panic right?

@evan-bradley what are your thoughts?

Reply:
Looking again at the draft implementation and how it hooks into the transformprocessor I see how this is enabling us to move the "Collection" idea from the processor into OTTL, which was our goal. I'm beginning to think it would be better, though, if we keep the Signal as part of the Collection, like the transformprocessor had it.  

We could potentially expose a generic solution in the future, but for the start I'm wondering if the API is safer if we expose a `MetricsParserCollection` that requires R to be a `consumer.Metrics` interface.

Reply:
We'd still have the panic opportunity in the happy-path return statment tho: `return convertedStatements.Interface().(R), nil`. There is no guarantee that `convertedStatements.Interface()` can be an R.

Reply:
> We could potentially expose a generic solution in the future, but for the start I'm wondering if the API is safer if we expose a MetricsParserCollection that requires R to be a consumer.Metrics interface and the same for Traces and Logs.

IMO, In terms of safety, having strict types would definitely make it harder to consumers misusing the API, but in general, what works for `consumer.X` should also work for any type of `R`.

We would also need to include the `go.opentelemetry.io/collector/consumer` dependency into the ottl package. But it's true that testing would be easier considering we have a limited number of accepted types. 

> We'd still have the panic opportunity in the happy-path return statment tho: return convertedStatements.Interface().(R), nil. There is no guarantee that convertedStatements.Interface() can be an R.

I might be missing something, but it should be guaranteed (public API level) by the `WithParserCollectionContext` option (`ParserCollectionOption`), which requires both parser collection and converter to have the same R type.

Reply:
But does that guarantee that the `reflect.Value` returned from `statementsConverter.call` is of type R?

Reply:
> If I'm not missing anything, It should be enough to guarantee that all parser collection converters returns the same R type.

I agree with this.  I am more interested if, down in `ParseStatementsWithContext`  the value returned from `contextParser.statementsConverter.call` is guaranteed to be castable to `R`.

```go
        // Are we guaranteed that convertedStatements can be cast to an R?
        // I know the compiler will think we can cast this to R, but since it is 
        // a reflect.Value I don't think we have that guarantee. For example
        // if R is an int64, what happens here?
        convertedStatements, err := contextParser.statementsConverter.call(
		reflect.ValueOf(pc),
		contextParser.ottlParser,
		context,
		statements,
		parsedStatements,
	)

	if err != nil {
		return *new(R), err
	}

	return convertedStatements.Interface().(R), nil
```

Reply:
Oh wait I think I see it now. `ParsedStatementConverter` is a function that MUST return R. Thats the function that gets called here:

```go
contextParser.statementsConverter.call(
		reflect.ValueOf(pc),
		contextParser.ottlParser,
		context,
		statements,
		parsedStatements,
	)
```

So we're good. While it returns a reflect.Value, we KNOW that value is an R bc of the type of the `ParsedStatementConverter`.

Reply:
Assuming my understanding is correct, I am ok not introducing signal-specific stuff in the OTTL package. If needed we could add that to the ottlcontexts (if we wanted to hide the generics from the transformprocessor)

Reply:
Yes, that's correct! It's guaranteed by the public API. 

I've pushed the extra condition to avoid panicking with `nil` return (https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/36174/commits/013c309eada994ec3a2210f6e21e102f9990ef7d)